### PR TITLE
Feature: add advertise and publish to the extension api

### PR DESCRIPTION
**Pull request by defunctzombie**
## User-Facing Changes
Panels authored via the extension api are able to advertise datatypes on topics.
Panels authored via the extension api are able to publish messages on advertised topics.

### Advertise
The advertise method on the panel extension context allows a panel to indicate an intent to publish a specific datatype on a topic. The panel would need to call advertise before being able to publish on the topic.

```typescript
context.advertise(topic: string, datatype: string, options?: Record<string, unknown>);
```

Options are specific to the current data source. Some data sources make use of specific options while others do not.

### Unadvertise 
Remove advertising for a topic/datatype pair.

### Publish
The publish function publishes a message on a previously advertised topic. If the topic is not advertised or otherwise malformed, the function will throw.

```typescript
context.publish(topic: string, message: Record<string, unknown>);
```

This function is fire-and-forget.

## Description
The user facing changes capture the scope of the PR.

### CODE REVIEW

File: packages/studio-base/src/panels/Publish/PublishPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    // fixme - is the info enough to advertise or do we need the list of datatypes?
+    // actually need the message definition in full
+    const info = datatypeInfo?.get(selectedDatatype);
+    if (!info) {
+      return;
+    }
+
+    const topicName = selectedTopic;
+    context.advertise(topicName, selectedDatatype, info);
```
Comment (jtbandes) [inline packages/studio-base/src/panels/Publish/PublishPanel.tsx:151]: why do we need to pass in info if it just comes from datatypeInfo which itself comes from the context?

File: packages/studio/index.d.ts [hunk 2]
```diff
@@ -16,6 +16,14 @@ declare module "@foxglove/studio" {
     datatype: string;
   };
 
+  export type DatatypeInfo = {
+    /** The type of datatype info. This is specific to each data source. */
+    type: string;
+
+    /** Additional arguments based on the type */
+    [key: string]: unknown;
+  };
```
Comment (jtbandes) [inline packages/studio/index.d.ts:25]: `{ type: "ros1", datatypes: RosDatatypes } | {type: string; [key: string]: unknown}` ?

File: packages/studio-base/src/components/PanelExtensionAdapter.tsx [hunk 3]
```diff
@@ -287,11 +309,64 @@ function PanelExtensionAdapter(props: PanelExtensionAdapterProps): JSX.Element {
...
+        rosDatatypes[datatype] = datatypeInfo.detail as RosDatatype;
+
+        ctx.setPublishers(subscriberId, [
+          {
+            topic,
+            datatype,
+            datatypes: rosDatatypes,
+            advertiser: { type: "panel", name: subscriberId },
+          },
+        ]);
```
Comment (jtbandes) [inline packages/studio-base/src/components/PanelExtensionAdapter.tsx:335]: may need to be updated to handle multiple advertise calls per panel?

**Comment by defunctzombie** [comment]
I've removed the DatatypeInfo and reworking Publish panel to extensions api aspects of this PR to focus on advertise and publish apis only. Those will be enough to support the new control panel which I'll update on top of this api once we merge.

**Comment by esthersweon** [comment]
Friendly reminder to add `docs` label to PRs like this.

**Pull request merged**